### PR TITLE
Netty HTTP client and server instrumentation

### PR DESCRIPTION
**Pull request by tylerbenson**
Because the [channel attribute map](https://netty.io/wiki/new-and-noteworthy-in-4.0.html#attributemap) was only introduced in 4.0, it will require a different approach to instrument earlier versions.  Perhaps using a weak keyed map?

### CODE REVIEW

**Comment by tylerbenson** [comment]
@realark What do you think about this approach?

**Comment by realark** [comment]
This looks like a good approach. Is there any risk of an unexpected use of a pipeline to consider? For example, a pipeline which hooks up an http decoder but not an encoder.

**Comment by tylerbenson** [comment]
Yeah. That is one risk. Another is if someone adds and later removes it. Any ideas for how to handle those cases better or how common they are?

**Comment by tylerbenson** [comment]
@realark for this reason, I don't think it will be very easy to support netty <4: https://netty.io/wiki/new-and-noteworthy-in-4.0.html#attributemap.  I'm using that to store the scope.  If you think I should explore further, I would need to look into using a weak keyed map from the channel or something.

**Comment by tylerbenson** [comment]
(Still working on adding a version scan)

File: dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/server/HttpServerRequestTracingHandler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            .withTag(Tags.PEER_HOSTNAME.getKey(), remoteAddress.getHostName())
+            .withTag(Tags.PEER_PORT.getKey(), remoteAddress.getPort())
+            .withTag(Tags.HTTP_METHOD.getKey(), request.getMethod().name())
+            .withTag(Tags.HTTP_URL.getKey(), url)
+            .withTag(Tags.COMPONENT.getKey(), "netty")
+            .withTag(DDTags.SPAN_TYPE, DDSpanTypes.WEB_SERVLET)
+            .startActive(false);
+
+    if (scope instanceof TraceScope) {
+      ((TraceScope) scope).setAsyncPropagation(true);
```
Comment (realark) [inline dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/server/HttpServerRequestTracingHandler.java]: I think we'll also want to disabled the async propagation in `HttpServerResponseTracingHandler`

Comment (tylerbenson) [reply dd-java-agent/instrumentation/netty-4.0/src/main/java/datadog/trace/instrumentation/netty40/server/HttpServerRequestTracingHandler.java]: Are you sure?  I close the scope on line 68.  I'm only passing the span to the `HttpServerResponseTracingHandler`, so I wouldn't be able to disable it unless I change what I'm passing.

**Pull request merged**